### ISSUE TITLE
[Client] Weird hang ups, lots of DB errors, GUI completely irresponsive until taskkill and crashing

### ISSUE THREAD

**Issue by HASJ** [hydrusnetwork/hydrus#96]
[Log file hosted @ google drive](https://drive.google.com/file/d/0BwLhfN-Yum0heW9QWHc4aHR3MkU/view?usp=sharing)
**Weird hang ups**: happens more frequently when I'm downloading from a booru. Most likely related to tagging but there's no affect to the PC's overall performance.
Hydrus just hangs for a few seconds and comes back. However, sometimes, for no good reason when I have a window within Hydrus opened (only observed when using Manage tags, tag siblings, tag parents; did not test with other windows), the program hangs and when it comes back the window that was on foreground falls behind Hydrus and then I have to alt+tab into Hydrus again for it to come out.

**DB errors**: the log is filled with them, I dunno what they mean. Is my DB corrupted? Already?
most common is this one:
```
Database Error!
'utf8' codec can't decode byte 0xe3 in position 1265: invalid continuation byte
Unknown Caller, probably GUI.
Traceback (most recent call last):

  File "C:\code\Hydrus\build\client\out00-PYZ.pyz\include.HydrusDB", line 198, in _ProcessJob

  File "C:\code\Hydrus\build\client\out00-PYZ.pyz\include.ClientDB", line 6689, in _Write

  File "C:\code\Hydrus\build\client\out00-PYZ.pyz\include.ClientDB", line 5050, in _SetJSONDump

  File "C:\code\Hydrus\build\client\out00-PYZ.pyz\json", line 243, in dumps

  File "C:\code\Hydrus\build\client\out00-PYZ.pyz\json.encoder", line 207, in encode

  File "C:\code\Hydrus\build\client\out00-PYZ.pyz\json.encoder", line 270, in iterencode

UnicodeDecodeError: 'utf8' codec can't decode byte 0xe3 in position 1265: invalid continuation byte
```

**GUI "irresponsiveness"**: It keeps on running normally but I can't do anything, not even normally close it. This is the second time this happened. Around line 50000+- on the log.
I suppose the "Unknown Caller, probably GUI" caller traceback has something to do with it.
**EDIT**: BTW, you know when you click a background window and Windows play that sound telling you that something else is on foreground? That's what that happens when the GUI does this hang and I click on it. Weirdly enough, the traceback messages are still clickable and can be closed as normal, with a right click.
**EDIT 2**:This time, I did not kill the program, went to sleep and a new thing has happened. The GUI stopped altogether, the rest of the program seems to have frozen as well, even the previously working traceback messages went to shit.
Clicked around the window and eventually Windows gave me the "stopped responding, kill or wait" message. Chose to kill it and even then Hydrus did not die. Windows has never done that. Waited for 5 minutes and Hydrus' icon was still on the taskbar, activated the window, clicked on it and it vanished. WTF. Seriously, **What the fuck**?

**Crashing**: There have been 4 crashes but I can't identify where they happened in the log, the lack of a clock in the log makes it hard to say.
I can say to you however that these crashes happened after one of the weird hang ups and a booru/files tab was active. All I did was click wrong while Hydrus hanged, when it did came back, it'd do my last action (either scroll or select a file) then close or just instantly vanish. Taskkill like.

**EDIT 3**: Guess I should say I'm running this on a toaster. My main notebook is on repair. Also, Windows 10, but I don't think Windows version matter much after W7.

**Comment by HASJ** [comment]
I'm starting to think this is related to how large my download tabs get.
I currently have two booru download tabs opened. One has 4072/24728 images and another has 4526/16301 images.
There's no way to clear out the already downloaded thumbnails.
A lot of these images have been already downloaded, many within these tabs, but Hydrus doesn't seem to sort that out by filtering duplicated links or something (this would save a great deal of time and bandwidth).

**Comment by hydrusnetwork** [comment]
Thank you for these reports and the log. I am sorry for the delay to getting to this--I am prioritising catching up on my github queue now.

Unfortunately, the program is very CPU and HDD hungry and python, the language I use to write it, is not very good at doing more than one thing at once. When the client imports a file, a lot of CPU time needs to be used, and for that period, the gui will be unresponsive. For most computers, this is not a long delay, so it is not very noticeable, but on slower machines, it can be a problem. Having pages with thousands of files open can compound it as lists of tens of thousands of tags are being recalculated on every new file.

I am always trying to optimise the program to eliminate bottlenecks like you are seeing, and reports like your log are helpful for that. Whenever you encounter anything that is particularly and reliably slow, you can try _help->debug->db profile mode_ while it is happening, which will spam detailed code profile information to the log. Shut the profile mode off after a while, close the client to flush the log, and then send me the profile tables at the end of the log. I can then see which lines are running so slow and work on optimising those areas.

I am not sure what those unicode errors are from. Do they happen regularly, say about every five minutes? If so, please try closing all pages, shutting the client down, and restarting. Let me know if that gets rid of them, and if they come back. I will improve that specific error for the next version to report more information.

For a slow computer, I recommend not having more than 1000 files open in any single page, not to run more than one search or download at once, and go to _file->options->maintenance and processing_ and uncheck "Run jobs when the client is not busy?". Otherwise, if too many things are running at once, the gui gets choked and can't respond to your mouse clicks. Also, it won't respond to Windows's requests to refresh or focus and so on, which is usually the cause of the odd window order and alt-tab behaviour.

I suspect the EDIT 2 hang was the client running a big job in the background and not being able to cancel it fast enough to return gui control to you. If you do move your big jobs to shutdown time, make sure you shut the client at least once a week so it can finish whatever has piled up.

Running a vacuum (_database->maintenance->vacuum_) after a lot of recent changes can also speed things up a bit (it basically defrags the db file), although it takes a few minutes to complete. Make sure you defrag your hard drive normally as well.

To remove thumbnails from a page, select them and right click for their menu. _remove selected_ will remove them from the view. You can close the whole page by middle clicking on its tab, or hitting Ctrl+W.

You have several "429 Too many requests - please slow down..." errors from some booru in your log, so if you haven't changed it yet, try increasing the 'polite wait' option in _options->downloading_.

Please let me know whether slowing down and reducing the size of your queries improves your gui responsiveness, and if you notice doing anything in particular is particularly detrimental.

**Comment by HASJ** [comment]
Can't give you feedback right now because of #122.

**Comment by HASJ** [comment]
Okay. Let's go.
Every single hang up and apparently even the DB decode error went away after doing:

>To remove thumbnails from a page, select them and right click for their menu. remove selected will remove them from the view.

The download queries are still large but removing the literal tens of thousands of thumbs the client had to cache from it made it work like the first time.

>Running a vacuum (database->maintenance->vacuum) after a lot of recent changes can also speed things up a bit (it basically defrags the db file), although it takes a few minutes to complete.

I let the automatic maintenance run whenever the client asks me too, last time was friday.

For now, I'm gonna let the client run during the night and downloading the files to see if the DB decode error shows up again, then delete some files and repeat the process periodically cleaning the tab with 'remove selected'.
I saw no detriment to the display and detection of the unable to decode aside from the annoyance, as in, I do not know what the client is actually failing to do, other than it could possibly be the cause of the crashes I was experiencing.

Now I'll run these tests and report to you in a couple of days.

**Comment by HASJ** [comment]
@hydrusnetwork 
Hey there.

Maybe you fixed the bug between v183 and v185 or it was being triggered by having 20k thumbs in the download tabs but it is gone now.
The client has been running for 48 hours. Running time isn't a factor
It downloaded 8k files. Number of files also isn't a factor.
In the first 24 hours I periodically removed the thumbs from the download tab. Nothing happened except the client running blazing fast so I can say the client (or my computer) can't handle having 20k cached thumbs.
That also influenced the boot time. At first I thought it was just because my db has grown into a 60gb monster but it was all because of the cached thumbs. Maybe you could display a warning for this?

In the last 24 hours I just let Hydrus do it's thing and didn't interfere. Nothing either. 2k files in the tab and no impact on performance, no small hang ups, no irresponsiveness, no crashes, etc.

There are 14k files to download still (yeah, I have a bit of an OCD problem) and I going to let Hydrus cache all of that and see if the decode error pops up again. Performance will also take a hit so I'll try to pinpoint when it starts happening so you can do a warning or whatever to mitigate it.

**Comment by hydrusnetwork** [comment]
Now I have fixed #125 , I am going to chalk this up to a combination of that error and having lots of thumbnails open at once. I recommend keeping your download queues to <1000 if you can. If you find you just want to get more and more, remember that getting the files is the easy bit--filtering and tagging is what takes time. Try filtering all those files, and you will gain a new appreciation for being selective! (I have had a non-empty inbox about 21 months now--I have done about 360,000 total and still have ~42,000 to go, with about a thousand more added every week).

Anyway, thank you for the information here. I will create a job to look at why having lots of thumbnails around slows the gui so much. I am glad your client is running better--please keep me updated on how it continues in future.
